### ISSUE TITLE
Vanilla Spiking RNNs

### ISSUE THREAD

**Issue by npvoid** [fangwei123456/spikingjelly#91]
I was interested in using a vanilla RNN with LIF neurons. However, it seems the current implementation of `SpikingVanillaRNNCell`  does implement any membrane dynamics.

https://github.com/fangwei123456/spikingjelly/blob/42e489f33b637f9d47945e9d993cfb342baaa4c0/spikingjelly/clock_driven/rnn.py#L829-L832

Is there a workaround for this? Basically what I would like to use is something like `neuron.LIFNode`  but with the spiking outputs of that layer being also feed back into itself.

**Comment by fangwei123456** [comment]
Hi, I write spiking RNNs according to **Long Short-Term Memory Spiking Networks and Their Applications**(https://arxiv.org/abs/2007.04779). This paper only proposes spiking LSTM, but I also implement some other spiking RNNs according to its method.

In my opinion, the only difference between spiking RNNs and RNNs is that spiking RNNs use heavisie activation function, while RNNs use sigmoid and tanh activation function. So, the API of spiking RNNs are designed to mimic RNNs in pytorch, which is different with the API of spiking neurons in SpikingJelly. The hidden states are stored in spiking neuron itself, but as input and output for spiking RNNs.

I guess you want to use spiking RNNs as spiking neurons that store hidden states in RNNs itself?

**Comment by npvoid** [comment]
I agree that the main difference is the heaviside activation. However, while that would give you a spiking RNN, it wouldn't be a LIF layer since the reset and leaky dynamics aren't there. I think the current implementation is like an IF with no reset. 

I believe the workaround is simple but a bit cumbersome at the same time since you have to manually define the recurrent weights and manipulate shapes a bit

```
import torch
import torch.nn as nn
from spikingjelly.clock_driven import neuron

T, input_size, output_size = 50, 100, 10
tau = 100.
W_r = torch.randn(output_size, output_size)
W_r[torch.eye(output_size).byte()] = 0.

net1 = nn.Sequential(
    nn.Linear(input_size+output_size, output_size, bias=False),
    neuron.LIFNode(tau=tau)
    )

in_spks = (torch.rand(T, input_size)>0.95).float()
out_spks = torch.zeros(output_size)
for t in range(T):
    out_ = torch.mm(W_r, out_spks.view(-1, 1)).flatten()
    out_spks = net1(torch.cat((in_spks[t, :], out_))) 
```

**Comment by fangwei123456** [comment]
You mean you need a neuron, whose input is not only from the last layer, but also from its output spikes at last time-step?

**Comment by npvoid** [comment]
Yes, exactly its inputs are from the previous layer (as usual) as well  as its output spikes. This has been extensively used (for example here fig. 6 https://www.biorxiv.org/content/10.1101/2020.06.29.176925v1)

**Comment by fangwei123456** [comment]
You can try this example:
```
from spikingjelly.clock_driven import neuron, functional, base
import torch
import torch.nn as nn
import torch.nn.functional as F

class SelfConnectedContainer(base.MemoryModule):
    def __init__(self, sub_module: nn.Module, in_features: int, out_features: int, bias: bool = True) -> None:
        super(SelfConnectedContainer, self).__init__()
        self.fc = nn.Linear(in_features + out_features, out_features, bias)
        self.sub_module = sub_module
        self.register_memory('out_spike', None)

    def forward(self, x: torch.Tensor):
        if self.out_spike is None:
            if x.ndim == 2:
                self.out_spike = torch.zeros([x.shape[0], self.fc.out_features]).to(x)
            else:
                out_shape = [x.shape[0]]
                out_shape.extend(x.shape[1:-1])
                out_shape.append(self.fc.out_features)
                self.out_spike = torch.zeros(out_shape).to(x)
        x = torch.cat((x, self.out_spike), dim=-1)
        self.out_spike = self.sub_module(self.fc(x))
        return self.out_spike


# test codes
in_features = 4
out_features = 2
T = 8
N = 2
net = SelfConnectedContainer(neuron.LIFNode(), in_features, out_features)
print(net)
x = torch.rand([T, N, in_features])
for t in range(T):
    print(t, net(x[t]))

functional.reset_net(net)



```

**Comment by fangwei123456** [comment]
> You can try this example:
> 
> ```
> from spikingjelly.clock_driven import neuron, functional, base
> import torch
> import torch.nn as nn
> import torch.nn.functional as F
> 
> class SelfConnectedContainer(base.MemoryModule):
>     def __init__(self, sub_module: nn.Module, in_features: int, out_features: int, bias: bool = True) -> None:
>         super(SelfConnectedContainer, self).__init__()
>         self.fc = nn.Linear(in_features + out_features, out_features, bias)
>         self.sub_module = sub_module
>         self.register_memory('out_spike', None)
> 
>     def forward(self, x: torch.Tensor):
>         if self.out_spike is None:
>             if x.ndim == 2:
>                 self.out_spike = torch.zeros([x.shape[0], self.fc.out_features]).to(x)
>             else:
>                 out_shape = [x.shape[0]]
>                 out_shape.extend(x.shape[1:-1])
>                 out_shape.append(self.fc.out_features)
>                 self.out_spike = torch.zeros(out_shape).to(x)
>         x = torch.cat((x, self.out_spike), dim=-1)
>         self.out_spike = self.sub_module(self.fc(x))
>         return self.out_spike
> 
> 
> # test codes
> in_features = 4
> out_features = 2
> T = 8
> N = 2
> net = SelfConnectedContainer(neuron.LIFNode(), in_features, out_features)
> print(net)
> x = torch.rand([T, N, in_features])
> for t in range(T):
>     print(t, net(x[t]))
> 
> functional.reset_net(net)
> ```

This implement has a bug. I fix it in https://github.com/fangwei123456/spikingjelly/commit/8f170585b72791cd56706f8fa7558837778f4915 .
